client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -144,13 +155,38 @@
         List<ByteBuffer> data = events.stream().map(serializer::serialize).collect(Collectors.toList());
         CompletableFuture<Void> ackFuture = new CompletableFuture<Void>();
         synchronized (writeFlushLock) {
-            synchronized (writeSealLock) {
-                SegmentOutputStream segmentWriter = getSegmentWriter(routingKey);
-                segmentWriter.write(PendingEvent.withHeader(routingKey, data, ackFuture));
+            if (data.stream().mapToInt(m -> m.remaining()).sum() > Serializer.MAX_EVENT_SIZE) {
+                writeLargeEvent(routingKey, data, ackFuture);
+            } else {
+                synchronized (writeSealLock) {
+                    SegmentOutputStream segmentWriter = getSegmentWriter(routingKey);
+                    segmentWriter.write(PendingEvent.withHeader(routingKey, data, ackFuture));
+                }
             }
         }
         return ackFuture;
     }
+    
+    @GuardedBy("writeFlushLock")
+    private void writeLargeEvent(String routingKey, List<ByteBuffer> events, CompletableFuture<Void> ackFuture) {
```

Comment:
Does anything actually complete `ackFuture`, besides in the `AuthenticationException` catch?

Reply:
Good catch.